tests/phpunit/tests/blocks/blockHooks.php

```diff
@@ -8,26 +8,76 @@
  * @since 6.4.0
  *
  * @group blocks
+ * @group block-hooks
  */
 class Tests_Blocks_BlockHooks extends WP_UnitTestCase {
 
+	/**
+	 * Registered block names.
+	 *
+	 * @var string[]
+	 */
+	private $registered_block_names = array();
+
+	/**
+	 * Registered pattern names.
+	 *
+	 * @var string[]
+	 */
+	private $registered_pattern_names = array();
+
 	/**
 	 * Tear down after each test.
 	 *
 	 * @since 6.4.0
 	 */
 	public function tear_down() {
-		$registry = WP_Block_Type_Registry::get_instance();
+		while ( ! empty( $this->registered_block_names ) ) {
+			$block_name = array_pop( $this->registered_block_names );
+			unregister_block_type( $block_name );
+		}
 
-		foreach ( array( 'tests/my-block', 'tests/my-container-block' ) as $block_name ) {
-			if ( $registry->is_registered( $block_name ) ) {
-				$registry->unregister( $block_name );
-			}
+		while ( ! empty( $this->registered_pattern_names ) ) {
+			$block_name = array_pop( $this->registered_pattern_names );
+			unregister_block_pattern( $block_name );
 		}
 
 		parent::tear_down();
 	}
 
+	/**
+	 * @see register_block_type()
+	 */
+	private function register_block_type( $block_type, $args = array() ) {
+		$result                         = register_block_type( $block_type, $args );
+		$this->registered_block_names[] = $result->name;
+
+		return $result;
+	}
```

Comment:
One alternative to "shadowing" `register_block_type` that I've seen elsewhere is using a private instance of the block type registry ([e.g.](https://github.com/WordPress/wordpress-develop/blob/f7d73158acd8bef9c0d98316c1cd5ecf4cd1a636/tests/phpunit/tests/blocks/wpBlockTypeRegistry.php#L13-L19)). (That might make setup and teardown a bit cleaner.)

Reply:
I shamelessly stole this code from the test suite covering Block Context:

https://github.com/WordPress/wordpress-develop/blob/f7d73158acd8bef9c0d98316c1cd5ecf4cd1a636/tests/phpunit/tests/blocks/context.php#L37-L47

https://github.com/WordPress/wordpress-develop/blob/f7d73158acd8bef9c0d98316c1cd5ecf4cd1a636/tests/phpunit/tests/blocks/context.php#L49-L66

These blocks need to be in the global registry from `get_hooked_blocks` to work correctly, and I only added the logic to track registered blocks to unregister them later.

A simpler alternative might be to iterate over all registered blocks and deregister these starting with `tests/` 🤔 What do you think? It could also work with patterns starting with the name of the test theme.

Reply:
I tied a revised approach with https://github.com/WordPress/wordpress-develop/pull/5340/commits/cd6f58204d2bbde67804f7eded0fe776b4a50bed. It's all contained in the `tear_down` now.